### PR TITLE
testing with mocking

### PR DESCRIPTION
**Pull request by fpolizzi**

### CODE REVIEW

File: src/test/java/com/fpolizzi/booking/CarBookingServiceMockitoTest.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        Car available = car("ABC123", false);
+        when(carService.getAllCars()).thenReturn(List.of(available));
+        when(carBookingDao.getCarBookings()).thenReturn(List.of());
+        when(carService.getCar("ABC123")).thenReturn(Optional.of(available));
+
+        UUID bookingId = underTest.bookCar(user(UUID.randomUUID()), "ABC123");
+
+        assertThat(bookingId).isNotNull();
+        verify(carBookingDao).saveBooking(any(CarBooking.class));
+    }
```
Comment (Dancan254) [inline src/test/java/com/fpolizzi/booking/CarBookingServiceMockitoTest.java:65]: Consider capturing the `CarBooking` with ArgumentCaptor and asserting the saved booking actually contains the expected user and car. 
Right now the verify(carBookingDao).saveBooking(any(CarBooking.class)) only proves that something was saved, not that it was correct.
Also assert that the returned bookingId matches saved.getId().
You can watch [this](https://www.youtube.com/watch?v=vDnwCQLqb1Q) to explain the [Argument Captor](https://site.mockito.org/javadoc/current/org/mockito/ArgumentCaptor.html)

File: src/test/java/com/fpolizzi/booking/CarBookingServiceMockitoTest.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    @Test
+    void bookCar_throwsException_whenNoAvailableCars() {
```
Comment (Dancan254) [inline src/test/java/com/fpolizzi/booking/CarBookingServiceMockitoTest.java]: Rename to ....whenAllCarsAreAlreadyBooked coz the car exists, it's just not available.

File: src/test/java/com/fpolizzi/car/CarServiceMockitoTest.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        List<Car> cars = List.of(
+                new Car("ABC123", BigDecimal.valueOf(50), CarBrand.AUDI, false),
+                new Car("XYZ789", BigDecimal.valueOf(80), CarBrand.BYD, true),
+                new Car("DEF456", BigDecimal.valueOf(60), CarBrand.FIAT, false)
+        );
+        when(carDao.getAllCars()).thenReturn(cars);
+
+        List<Car> result = underTest.getAllCars();
+
+        assertThat(result).hasSize(3);
```
Comment (Dancan254) [inline src/test/java/com/fpolizzi/car/CarServiceMockitoTest.java]: `hasSize(3)` only checks count, not content. Use `containsExactlyInAnyOrder(...)` to verify the actual cars returned, consistent with how you assert in the other tests below like in `getAllElectricCars_shouldReturnOnlyElectricCars()`.

File: src/test/java/com/fpolizzi/user/UserServiceMockitoTest.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    void getUsers_shouldReturnAllUsers() {
+        List<User> users = List.of(
+                new User(UUID.randomUUID(), "Alice", "Smith"),
+                new User(UUID.randomUUID(), "Bob", "Jones")
+        );
+        when(userDao.getUsers()).thenReturn(users);
+
+        List<User> result = underTest.getUsers();
+
+        assertThat(result).hasSize(2);
```
Comment (Dancan254) [inline src/test/java/com/fpolizzi/user/UserServiceMockitoTest.java]: Same issue as `CarServiceMockitoTest`:  `hasSize(2)` doesn't verify the actual users returned. consider adding this too `containsExactlyInAnyOrder(...)`.

File: src/test/java/com/fpolizzi/user/UserServiceMockitoTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    @Test
+    void getUserById_shouldReturnNull_whenListIsEmpty() {
+        when(userDao.getUsers()).thenReturn(Collections.emptyList());
+
+        User result = underTest.getUserById(UUID.randomUUID());
+
+        assertThat(result).isNull();
+    }
```
Comment (Dancan254) [inline src/test/java/com/fpolizzi/user/UserServiceMockitoTest.java:75]: `getUserById` returns null you can consider returning `Optional<User>` instead to make the "not found" case explicit and avoid Null Pointer Exception(NPEs). See how you handles this in `CarServiceMockitoTest` : `getCar` method .

**Review by Dancan254 — COMMENTED** [review]
Overall, solid test structure and good coverage across and edge cases.
 strengthen assertions beyond `hasSize()` to verify actual content that is returned, and prefer Optional over null returns you've handled both well in places, just apply them consistently across all services.

File: src/test/java/com/fpolizzi/booking/CarBookingServiceMockitoTest.java (hunk 1 above)
Comment (fpolizzi) [reply src/test/java/com/fpolizzi/booking/CarBookingServiceMockitoTest.java:65]: Thank you very much Ian for giving me very valuable insights and improvements!

File: src/main/java/com/fpolizzi/Main.java [hunk 6]
```diff
@@ -97,7 +97,7 @@ private static void displayAllUserBookedCars(UserService userService,
         System.out.println("➡️ select user id");
         String userId = scanner.nextLine();
 
-        User user = userService.getUserById(UUID.fromString(userId));
+        User user = userService.getUserById(UUID.fromString(userId)).orElse(null);
```
Comment (Dancan254) [inline src/main/java/com/fpolizzi/Main.java]: You're unwrapping `Optional<User>` back to `null` immediately with `.orElse(null)`, which brings back the same NPE risk Optional was meant to eliminate. Consider using `.ifPresentOrElse(...)` or an if `(result.isEmpty())` early return pattern instead.

File: src/main/java/com/fpolizzi/Main.java [hunk 7]
```diff
@@ -125,7 +125,7 @@ private static void bookCar(UserService userService, CarBookingService carBookin
         String userId = scanner.nextLine();
 
         try {
-            User user = userService.getUserById(UUID.fromString(userId));
+            User user = userService.getUserById(UUID.fromString(userId)).orElse(null);
```
Comment (Dancan254) [inline src/main/java/com/fpolizzi/Main.java]: The comment above applies to this as well.

**Review by Dancan254 — COMMENTED** [review]
Just those two related fixes and we're good

**Comment by fpolizzi** [comment]
> Just those two related fixes and we're good

How could I missed that! Hope I got it now finally.

Thank you again a lot Ian.
